spec/models/account_spec.rb

```diff
@@ -63,45 +131,150 @@
     let(:expires_at) { 1.day.from_now.to_i }
     let!(:account) { Account.create!(synced_id: 123, oauth_access_token: "token",
       oauth_refresh_token: "refresh_token", oauth_expires_at: expires_at) }
+    let(:oauth_client) { double }
+
+    before do
+      allow(account).to receive(:oauth_client).with(no_args).and_return(oauth_client)
+    end
 
     context "when the stored token is fresh" do
       it "returns the token" do
         expect(account.token).to be_a OAuth2::AccessToken
         expect(account.token.token).to eq "token"
       end
     end
-
+  
     context "when the stored token is expired" do
       # comparing rails version, the use_transactional_fixtures only works pre 5
       if Rails::VERSION::STRING.split(".").first.to_i >=5
         self.use_transactional_tests = false
       else
         self.use_transactional_fixtures = false
       end
-
+  
       let(:expires_at) { 1.day.ago.to_i.to_s }
       let(:new_expires_at) { 2.days.from_now.to_i.to_s }
-      let(:token) { double(expired?: true, refresh!: double(token: "refreshed_token",
-        refresh_token: "refreshed_refresh_token", expires_at: new_expires_at)) }
-      let(:client) { double }
-
+      let(:token) do
+        double(expired?: true, refresh!: double(token: "refreshed_token",
+          refresh_token: "refreshed_refresh_token", expires_at: new_expires_at))
+      end
+  
       before do
-        expect(BookingSync::Engine).to receive(:oauth_client) { client }
-        expect(OAuth2::AccessToken).to receive(:new).with(client, "token",
+        expect(OAuth2::AccessToken).to receive(:new).with(oauth_client, "token",
           refresh_token: "refresh_token", expires_at: expires_at) { token }
       end
-
+  
       after do
         Account.destroy_all
       end
-
+  
       it "refreshes the token" do
         expect(token).to receive(:refresh!)
         account.token
       end
     end
   end
 
+  describe "#application_token" do
+    let(:client_credentials) { double }
+    let(:oauth2_client) { double }
+
+    context "without multi application support" do
+      let!(:account) { Account.create!(synced_id: 123) }
+
+      it "returns a client credential token setup without default params" do
+        expect(client_credentials).to receive(:get_token).and_return("client_credentials_token")
```

Comment:
wow, do we need all these mocks here? maybe extraction to some service and mocking just this service would help here?

Reply:
I suppose we could test the engine methods in isolation.